pydis_site/apps/api/tests/test_allowlists.py

```diff
@@ -0,0 +1,112 @@
+from django_hosts.resolvers import reverse
+
+from pydis_site.apps.api.models import AllowDenyList
+from pydis_site.apps.api.tests.base import APISubdomainTestCase
+
+URL = reverse('bot:allowdenylist-list', host='api')
+JPEG_ALLOWLIST = {
+    "type": 'FILE_FORMAT',
+    "allowed": True,
+    "content": ".jpeg",
+}
+PNG_ALLOWLIST = {
+    "type": 'FILE_FORMAT',
+    "allowed": True,
+    "content": ".png",
+}
+
+
+class UnauthenticatedTests(APISubdomainTestCase):
+    def setUp(self):
+        super().setUp()
+        self.client.force_authenticate(user=None)
+
+    def test_cannot_read_allowedlist_list(self):
+        response = self.client.get(URL)
+
+        self.assertEqual(response.status_code, 401)
+
+
+class EmptyDatabaseTests(APISubdomainTestCase):
+    def test_returns_empty_object(self):
+        response = self.client.get(URL)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json(), [])
+
+
+class FetchTests(APISubdomainTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        cls.jpeg_format = AllowDenyList.objects.create(**JPEG_ALLOWLIST)
+        cls.png_format = AllowDenyList.objects.create(**PNG_ALLOWLIST)
+
+    def test_returns_name_in_list(self):
+        response = self.client.get(URL)
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json()[0]["content"], self.jpeg_format.content)
+        self.assertEqual(response.json()[1]["content"], self.png_format.content)
+
+    def test_returns_single_item_by_id(self):
+        response = self.client.get(f'{URL}/{self.jpeg_format.id}')
+
+        self.assertEqual(response.status_code, 200)
+        self.assertEqual(response.json().get("content"), self.jpeg_format.content)
+
+    def test_returns_allow_deny_list_types(self):
+        response = self.client.get(f'{URL}/get_types')
+
+        self.assertEqual(response.status_code, 200)
+        for api_type, model_type in zip(response.json(), AllowDenyList.AllowDenyListType.choices):
+            self.assertEquals(api_type[0], model_type[0])
+            self.assertEquals(api_type[1], model_type[1])
+
+
+class CreationTests(APISubdomainTestCase):
+    def test_returns_400_for_missing_params(self):
+        no_type_json = {
+            "allowed": True,
+            "content": ".jpeg"
+        }
+        no_allowed_json = {
+            "type": "FILE_FORMAT",
+            "content": ".jpeg"
+        }
+        no_content_json = {
+            "allowed": True,
+            "type": "FILE_FORMAT"
+        }
+        cases = [{}, no_type_json, no_allowed_json, no_content_json]
+
+        for case in cases:
+            response = self.client.post(URL, data=case)
+            self.assertEqual(response.status_code, 400)
```

Comment:
This should use a subtest. I assume Django supports those.

Reply:
resolved in 99fc3cd